test: add wait_for_cookie_credentials() to framework for rpcwait tests

jonatack opened the pull request:
This PR adds a `wait_for_cookie_credentials()` method to the test framework and calls it before the `-rpcwait` tests, to avoid an intermittent race condition on the CI run with Valgrind where the cookie file isn't written yet when the CLI call with `-rpcwait` arrives to `get_auth_cookie()`.

To reproduce/test, build with
```diff
diff --git a/src/httprpc.cpp b/src/httprpc.cpp
index 60c4d06f12..3dd06c4758 100644
--- a/src/httprpc.cpp
+++ b/src/httprpc.cpp
@@ -291,6 +291,7 @@ static bool InitRPCAuthentication()
 bool StartHTTPRPC()
 {
     LogPrint(BCLog::RPC, "Starting HTTP RPC server\n");
+    UninterruptibleSleep(std::chrono::seconds{11});
     if (!InitRPCAuthentication())
```
then run the test normally and with valgrind
```
test/functional/interface_bitcoin_cli.py -l debug
valgrind test/functional/interface_bitcoin_cli.py -l debug
```
Thanks to Marco Falke for all the help. Closes #18684.

jonatack committed:
test: fix intermittent race condition in interface_bitcoin_cli.py

by calling wait_for_cookie_credentials() to ensure the cookie file is written
and auth credentials available for testing the CLI -rpcwait option before the
RPC connection is up.

MarcoFalke commented:
The error was "  Called Process failed with 'error: Could not locate RPC credentials. No authentication cookie could be found, and RPC password is not set.  See -rpcpassword and -stdinrpcpass.  Configuration file: (/home/travis/build/bitcoin/bitcoin/ci/scratch/test_runner/test_runner_₿_🏃_20200417_132645/interface_bitcoin_cli_109/node0/bitcoin.conf) "

jonatack commented:
Yes. I'm still looking at it, but since valgrind runs don't work out-of-repo, I wanted to see if the valgrind CI passes with this change while looking further.

jonatack commented:
Am noting that no other tests use `get_auth_cookie`... seems undertested...

MarcoFalke commented:
`get_auth_cookie` is used by every single test. It reads the auth cookie from disk, though it fails when the file was not yet created.

jonatack committed:
test: add wait_for_cookie_credentials() to test framework

to be able to ensure the cookie file is written and auth credentials available
when testing CLI/RPC commands before the RPC connection is up.

MarcoFalke commented:
Btw, this happens on cirrus reliably: https://cirrus-ci.com/task/6597545033990144?command=functional_test#L1053

MarcoFalke commented:
Or, if you want to test locally:

```diff
diff --git a/src/httprpc.cpp b/src/httprpc.cpp
index 60c4d06f12..3dd06c4758 100644
--- a/src/httprpc.cpp
+++ b/src/httprpc.cpp
@@ -291,6 +291,7 @@ static bool InitRPCAuthentication()
 bool StartHTTPRPC()
 {
     LogPrint(BCLog::RPC, "Starting HTTP RPC server\n");
+    UninterruptibleSleep(std::chrono::seconds{11});
     if (!InitRPCAuthentication())
         return false;

jonatack commented:
Thanks! Trying that now.

jonatack commented:
Thanks @MarcoFalke for your suggestions! Being able to reproduce makes all the difference. I updated the PR title and description and pushed a fix that seems to work.

jonatack commented:
I see that the CI for the latest merge on master (bump wallet_import timeout) failed for this same reason: https://travis-ci.org/github/bitcoin/bitcoin/jobs/676639823

jonatack commented on test/functional/test_framework/util.py:
```diff
@@ -354,7 +354,7 @@ def get_auth_cookie(datadir, chain):
     except OSError:
         pass
-    if user is None or password is None:
+    if (user is None or password is None) and wait_for_cookie is False:
```
Thanks for reviewing and your help. If I understand your comment correctly, I modified the conditional here not to change the interface, but to not raise when we're waiting for the cookie auth credentials. When they do become available, it still returns the auth pair which we're waiting for in `wait_for_cookie_credentials()`.

jonatack replied:
E.g. since `wait_for_cookie` is false by default, there is no change in behavior except when it's called by `wait_for_cookie_credentials`.

MarcoFalke replied:
Yeah, I don't understand why `wait_for_cookie` is needed? Just because it is turned off by default doesn't mean that it doesn't complicate the interface

jonatack replied:
Hm, maybe it isn't. I modified that line before changing the conditional on line 340, because I was seeing empty creds. Good question. Re-checking.

MarcoFalke replied:
They should never be empty. 

Either both are a string of length of at least 1 or a ValueError is thrown

jonatack replied:
It seems needed... I'll look more into why tomorrow, since it's getting late. Would you prefer a separate function to not change `get_auth_cookie`?

jonatack replied:
I was seeing empty values, which is why I had to check for them at https://github.com/bitcoin/bitcoin/pull/18691/files#diff-86294e5ae5283eebdd9f98d79007a0e1R257

jonatack replied:
I'll look further into what and why.

jonatack replied:
Ok, I think I see what you mean.

jonatack replied:
Good catch, fixed.

jonatack commented:
Simplified.

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -243,6 +244,20 @@ def wait_for_rpc_connection(self):
+            except ValueError:
+                pass
+            time.sleep(1.0 / poll_per_s)
+        self._raise_assertion_error("Unable to retrieve cookie credentials")
```
nit: Could rebase and adjust the error to mention the timeout (like the one above, which you can see after rebase)

jonatack replied:
good idea; done

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -243,6 +244,20 @@ def wait_for_rpc_connection(self):
+        for _ in range(poll_per_s * self.rpc_timeout):
+            try:
+                user, passphrase = get_auth_cookie(self.datadir, self.chain)
+                self.log.debug("Cookie credentials retrieved: user={}, passphrase={}".format(user, passphrase))
```
nit: the values aren't needed, only their presence, so no need to log them?

jonatack replied:
done

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -243,6 +244,20 @@ def wait_for_rpc_connection(self):
+                user, passphrase = get_auth_cookie(self.datadir, self.chain)
+                self.log.debug("Cookie credentials retrieved: user={}, passphrase={}".format(user, passphrase))
+                return
+            except ValueError:
```
nit: In a future cleanup the three occurrences of `ValueError` can be changed to a self-made error `RpcMissingCred` (or so). Not asking you to do this here.

jonatack replied:
Good idea; will look for a follow-up

jonatack replied:
> nit: In a future cleanup the three occurrences of `ValueError` can be changed to a self-made error `RpcMissingCred` (or so). Not asking you to do this here.

Good idea, was also just noticing the three occurrences of "No RPC credentials", etc. Will look at a follow-up.

maflcko reviewed (approved):
ACK 83b037921fac34d4c85ce6a2e137407596fed91e

Happy to merge with or without feedback addressed. Just let me know what you prefer.

jonatack commented:
Took the suggestions, and also updated two cli calls in the test so the debug log shows:
```diff
- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -getinfo

- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -version
```

jonatack commented:
Rebased and added a docstring to `wait_for_cookie_credentials()`.

MarcoFalke commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -76,20 +75,17 @@ def run_test(self):
         self.log.info("Test -version with node stopped")
-        cli_response = self.nodes[0].cli("-version").send_cli()
+        self.stop_node(0)
+        cli_response = self.nodes[0].cli().send_cli('-version')
```
Why is this changed?

MarcoFalke replied:
-version is not an rpc command, but a command line option

MarcoFalke commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -52,7 +51,7 @@ def run_test(self):
         if self.is_wallet_compiled():
             self.nodes[0].encryptwallet(password)
-        cli_get_info = self.nodes[0].cli('-getinfo').send_cli()
+        cli_get_info = self.nodes[0].cli().send_cli('-getinfo')
```
same here

jonatack replied:
Changed to improve the debug log output:
```diff
- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -getinfo

- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -version
```
Can revert in #18594 that also touches the testfile.

jonatack replied:
I changed it to improve the debug log output:
```diff
- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -getinfo

- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -version
```
I can revert it in #18594 that also touches the testfile or look at improving the logging output for these cases.

MarcoFalke replied:
That should be fixed by logging the options, not passing the command line option as the rpc command

jonatack replied:
> That should be fixed by logging the options, not passing the command line option as the rpc command

Done in #18712.
